### PR TITLE
Add Singulars multivariate factorization to Singular.jl

### PR DESCRIPTION
**Pull request by raulepure**
Added: - Singular's factorization for QQ, ZZ and Fp.
            - Singular's squarefree factorization for QQ and Fp

### CODE REVIEW

File: src/poly/poly.jl [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
        derivative, div, divides, direm, evaluate, exponent, exponent!,
-       exponent_vectors, finish, gen, has_global_ordering,
+       exponent_vectors, factor, finish, gen, has_global_ordering,
        inflate, isgen,
        ismonomial, isterm, jacobi, jet, lc, lt, lm, lead_exponent,
        monomials, MPolyBuildCtx,
        nvars, ordering, @PolynomialRing, primpart,
-       push_term!, remove, sort_terms!, symbols, terms,
-       total_degree, valuation, var_index, vars
+       push_term!, remove, sort_terms!, squarefree_factorization,
```
Comment (wbhart) [inline src/poly/poly.jl]: Can I suggest the name factor_squarefree. This seems to agree with the convention the Hecke people are using for various factoring routines. It also gets around needing to decide on British vs Americano spelling.

Comment (raulepure) [reply src/poly/poly.jl]: name is changed.

File: deps/src/rings.cpp [hunk 2]
```diff
@@ -250,14 +274,14 @@ void singular_define_rings(jlcxx::Module & Singular)
...
+   Singular.method("p_Jet",
+                   [](poly p, int i, ring r) {
+                       poly p_cp = p_Copy(p, r);
+                       return p_Jet(p_cp, i, r);
     });
-
-   Singular.method("p_Diff", [](poly p, int i, ring r) {
-        poly p_cp = p_Copy(p, r);
-        return p_Diff(p_cp, i, r);
+   Singular.method("p_Diff",
```
[the remark this answers is not in the document]
Comment (raulepure) [reply deps/src/rings.cpp]: I just rearranged the code so its structure is similar to (great) parts of the other code in rings.cpp

File: deps/src/rings.cpp [hunk 3]
```diff
@@ -250,14 +274,14 @@ void singular_define_rings(jlcxx::Module & Singular)
         nMapFunc map_func = reinterpret_cast<nMapFunc>(map_func_ptr);
         return p_PermPoly(p, perm, old_ring, new_ring, map_func);
     });
-
-   Singular.method("p_Jet", [](poly p, int i, ring r) {
-        poly p_cp = p_Copy(p, r);
-        return p_Jet(p_cp, i, r);
+   Singular.method("p_Jet",
```
[the remark this answers is not in the document]
Comment (raulepure) [reply deps/src/rings.cpp]: I just rearranged the code so its structure is similar to (great) parts of the other code in rings.cpp

File: src/poly/poly.jl [hunk 4]
```diff
@@ -618,6 +618,64 @@ f values")
...
+
+@doc Markdown.doc"""
+    factor(x::spoly)
+> Returns the factorization of $x$.
+"""
+function factor(x::spoly)
+  R = parent(x)
+
+  # Check if base ring is valid
+  if((R.base_ring != Singular.QQ) && (R.base_ring != Singular.ZZ) && (typeof(R.base_ring) != N_ZpField))
```
[the remark this answers is not in the document]
Comment (raulepure) [reply src/poly/poly.jl]: Done.

File: test/poly/spoly-test.jl [hunk 5]
```diff
@@ -455,7 +455,77 @@ function test_spoly_differential()
...
-   
+
+   println("PASS")
+end
+
+function test_spoly_factor()
+   print("spoly.test_spoly_factor...")
+
+   # Test over Singular.QQ
+   R1 , (x, y, z, w) = Singular.PolynomialRing(Singular.QQ, ["x", "y", "z", "w"]; ordering=:negdegrevlex)
```
[the remark this answers is not in the document]
Comment (raulepure) [reply test/poly/spoly-test.jl]: Done.

**Comment by wbhart** [comment]
Thanks Raul!

**Pull request merged**